esphome/config_helpers.py

```
@@ -38,7 +55,20 @@ def merge(old, new):
         elif isinstance(new, list):
             if not isinstance(old, list):
                 return new
-            return old + new
+            res = old.copy()
+            ids = {v[CONF_ID]: i for i, v in enumerate(res) if CONF_ID in v}
+            for v in new:
+                if CONF_ID in v:
+                    new_id = v[CONF_ID]
+                    if isinstance(new_id, Extend):
+                        new_id = new_id.value
+                    if new_id in ids:
```

It should raise an exception if the extended id doesn't exist

---

I'm leaving this up to the config validation code so that the user can get "nice" error messages. I know I could manually concoct one here, but it seemed like this was more validation's job.